### PR TITLE
[WIP] Add `ListDtype`

### LINKED ISSUE #5610: [DISCUSS] Python ListDtype
Edit: Please see https://github.com/rapidsai/cudf/pull/5629 for the initial implementation

As a first step towards enabling list types in `cudf`, we need to introduce a `ListDtype` in cuDF. What should this look like?

In Pandas, columns of lists have `'object'` dtype:

```python
In [14]: a = pd.Series([[1, 2], [1, 2, 3], [4,]])

In [15]: a
Out[15]:
0       [1, 2]
1    [1, 2, 3]
2          [4]
dtype: object

In [16]: a.dtype
Out[16]: dtype('O')
```

We can readily reject this approach for two reasons:

1. We currently use `'object'` as the dtype for string columns:

```python
In [2]: a = cudf.Series(['a', 'b', 'c'])

In [3]: a.dtype == 'object'
Out[3]: True
```

2. List columns in cudf are homogenous, in that all the values in a list column are of the same dtype. Our `dtype` object needs to be able to store that information somewhere .

PyArrow has `ListArrays`, which are of type `ListType`:

```python
In [6]: a = pa.ListArray.from_pandas(pd.Series([[1, 2], [1, 2, 3], [1,]]))

In [7]: a
Out[7]:
<pyarrow.lib.ListArray object at 0x7f473f75b830>
[
  [
    1,
    2
  ],
  [
    1,
    2,
    3
  ],
  [
    1
  ]
]

In [8]: a.type
Out[8]: ListType(list<item: int64>)

In [9]: a.type.value_type
Out[9]: DataType(int64)
```

Something like PyArrow's `ListType` is much better suited to our needs.

-------

Here's a skeleton `ListDtype` with some questions inline:

```python
class ListDtype(object): # should we inherit from Pandas ExtensionDtype?
    def __init__(self, value_type):  # should value_type have a default?
        self._value_type = value_type

    @property
    def value_type(self):
        return self._value_type
    
    @property
    def name(self):
        return "list" # should we choose something else?
   
    @classmethod
    def from_arrow(self, typ):
        pass

    def to_arrow(self):
        pass
    
    def __eq__(self, other):
        pass

    # What other methods, if any, could we define here?
 ```

cc: @brandon-b-miller as you've been thinking about dtypes more generally :)

### PR DESCRIPTION
**Pull request by shwina**
cc: @brandon-b-miller @kkraus14 Need to add tests but early feedback welcome

Closes #5610

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by brandon-b-miller** [comment]
One question I have generally is how should we handle casting to/from this? Should it just raise in every case or should we convert scalars into length-1 lists, or something else?

**Comment by shwina** [comment]
This is new ground. Seeing as it's not really possible in Pandas, and that PyArrow doesn't seem to support it, maybe we should throw unless someone asks for it?

```python
In [23]: a
Out[23]:
<pyarrow.lib.Int32Array object at 0x7eff1552df30>
[
  1,
  2,
  3
]

In [24]: a.type
Out[24]: DataType(int32)

In [25]: a.cast(pa.list_(pa.from_numpy_dtype('int32')))
---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
<ipython-input-25-b43305aa11db> in <module>
----> 1 a.cast(pa.list_(pa.from_numpy_dtype('int32')))

~/miniconda3/envs/cudf_dev/lib/python3.7/site-packages/pyarrow/array.pxi in pyarrow.lib.Array.cast()

~/miniconda3/envs/cudf_dev/lib/python3.7/site-packages/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowNotImplementedError: No cast implemented from int32 to list<item: int32>
```

**Comment by kkraus14** [comment]
+1 to throwing on casting. It's super ambiguous without having defined offsets. Especially when it comes to something like a StringColumn

File: python/cudf/cudf/core/dtypes.py [hunk 1]
```diff
@@ -103,3 +105,42 @@ def deserialize(cls, header, frames):
...
+        else:
+            return np.dtype(self._typ.value_type.to_pandas_dtype()).name
+
+    @property
+    def name(self):
+        return "list"
+
+    @property
+    def type(self):
+        return list
```
Comment (shwina) [inline python/cudf/cudf/core/dtypes.py:131]: Does it makes sense to do this? `type` should return something that can be used to create scalars of this dtype.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/dtypes.py:131]: Either this, or a numpy array that has the item type as its type, that way you'd convert scalars to single element numpy arrays. One vision I had for an eventual `cudf.dtype` was that all of our dtypes would use this api to produce `cudf.Scalar` objects that wrap the equivalent libcudf scalars.

Comment (kkraus14) [reply python/cudf/cudf/core/dtypes.py:131]: +1 to the vision of eventually being able to produce `cudf.Scalar` objects from doing something like `cudf.dtype("int64").type(5)`.

In the meantime, using `list` doesn't do a good job of capturing the dtype as a list can contain arbitrary Python objects whereas we're strongly typed.

**Comment by shwina** [comment]
Closing this in favour of just #5629 -- I found it difficult to manage the two branches separately when working locally, apologies.

**Pull request closed without merging**